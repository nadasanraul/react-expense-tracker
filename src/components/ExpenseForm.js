import React from 'react';
import moment from 'moment';
import 'react-dates/initialize';
import {SingleDatePicker} from 'react-dates';
import { create } from 'domain';

// // const date = new Date()
// const now = moment();
// console.log(now.format('Do MMM YYYY'));

export default class ExpenseForm extends React.Component{
    constructor(props){
        super(props);

        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? (props.expense.amount / 100).toString(): '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false,
            error: {
                descriptionError: '',
                amountError: ''
            },
            success: ''
        };
    }

    onDescriptionChange = e => {
        const description = e.target.value;
        this.setState(() => ({ description }));
    };

    onNoteChange = e => {
        const note = e.target.value;
        this.setState(() => ({ note }));
    };

    onAmountChange = e => {
        const amount = e.target.value;
        if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }));
        }
    };

    onDateChange = createdAt => {
        if(createdAt){
            this.setState(() => ({ createdAt }));
        }
    };

    onFocusChange = ({focused}) => {
        this.setState(() => ({calendarFocused: focused}));
    };

    onSubmit = (e) => {
        e.preventDefault();

        if(!this.state.description && !this.state.amount){
            //Set error state
            this.setState(() => ({
                error: {
                    descriptionError: 'Please provide a descrption',
                    amountError: 'Please Provide an amount'
                }
            }));
        } else if (!this.state.description){
            this.setState(() => ({
                error: {
                    descriptionError: 'Please provide a descrption',
                    amountError: ''
                }
            }));
        } else if (!this.state.amount){
            this.setState(() => ({
                error: {
                    descriptionError: '',
                    amountError: 'Please provide an amount'
                }
            }));
        } else {
            this.setState(() => ({
                error: {
                    descriptionError: '',
                    amountError: ''
                },
                success: 'You successfully added an expense'
            }));
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            });
            setTimeout(() => {
                this.setState(() => ({success: ''}));
            }, 3000);
        }
    };

    render(){
        return (
            <div>
                {
                    this.state.success &&
                    <div className="alert alert-success">
                        {this.state.success}
                    </div>
                }
                <form onSubmit={this.onSubmit}>
                    <div className="form-group">
                        <label>Description</label>
                        <input 
                            type="text" 
                            className={`form-control ${this.state.error.descriptionError && 'is-invalid'}`} 
                            autoFocus 
                            value={this.state.description}
                            onChange={this.onDescriptionChange}
                        />
                        {
                            this.state.error.descriptionError && 
                            <small 
                                className="invalid-feedback"
                            >
                                {this.state.error.descriptionError}
                            </small>
                        }
                    </div>
                    <div className="form-group has-danger">
                        <label>Amount</label>
                        <input 
                            type="text" 
                            className={`form-control ${this.state.error.amountError && 'is-invalid'}`}
                            value={this.state.amount}
                            onChange={this.onAmountChange}
                        />
                        {
                            this.state.error.amountError && 
                            <small 
                                className="invalid-feedback"
                            >
                                {this.state.error.amountError}
                            </small>
                        }
                    </div>
                    <div className="form-group">
                        <label>Date</label><br/>
                        <SingleDatePicker 
                            date={this.state.createdAt}
                            onDateChange={this.onDateChange}
                            focused={this.state.calendarFocused}
                            onFocusChange={this.onFocusChange}
                            numberOfMonths={1}
                            isOutsideRange={() => false}
                        />
                    </div>
                    <div className="form-group">
                        <label>Note</label>
                        <textarea 
                            placeholder="Add a note to your expenses(optional)"
                            className="form-control"
                            onChange={this.onNoteChange}
                            value={this.state.note}
                        >
                        </textarea>
                    </div>
                    <div className="form-group">
                        <button 
                            type="submit" 
                            className="btn btn-primary"
                        >
                            Add Expense
                        </button>
                    </div>
                </form>
            </div>
        )
    }
}