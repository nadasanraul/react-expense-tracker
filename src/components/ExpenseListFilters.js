import React from 'react';
import {connect} from 'react-redux';
import 'react-dates/initialize';
import {DateRangePicker} from 'react-dates';
import {setTextFilter, sortBy, setStartDate, setEndDate} from '../actions/filters';

export class ExpenseListFilters extends React.Component {
    state = {
        calendarFocused: null
    };

    onDatesChange = ({startDate, endDate}) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    };

    onFocusChange = (focused) => {
        this.setState(() => ({calendarFocused: focused}));
    };

    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value);
    };

    onSortChange = (e) => {
        this.props.sortBy(e.target.value);
    };

    render(){
        return (
            <div className="row mb-4">
                <div className="col-md-6">
                    <div className="form-group">
                        <input className="form-control" type="text" value={this.props.filters.text} onChange={this.onTextChange}/>
                    </div>
                </div>
                <div className="col-md-3">
                    <div className="form-group">
                        <select 
                            className="form-control"
                            onChange={this.onSortChange}
                            value={this.props.filters.sortBy}
                        >
                            <option value="date">Date</option>
                            <option value="amount">Amount</option>
                        </select>
                    </div>
                </div>
                <div className="col-md-6">
                    <DateRangePicker 
                        startDate={this.props.filters.startDate}
                        startDateId={this.props.filters.startDateId}
                        endDate={this.props.filters.endDate}
                        endDateId={this.props.filters.endDateId}
                        onDatesChange={this.onDatesChange}
                        focusedInput={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        showClearDates={true}
                        numberOfMonths={1}
                        isOutsideRange={() => false}
                    />
                </div>
            </div>
        );
    }
}

const mapStateToProps = state => ({filters: state.filters});

const mapDispatchToProps = dispatch => ({
    setTextFilter: text => dispatch(setTextFilter(text)),
    sortBy: value => dispatch(sortBy(value)),
    setStartDate: startDate => dispatch(setStartDate(startDate)),
    setEndDate: endDate => dispatch(setEndDate(endDate))
})

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);