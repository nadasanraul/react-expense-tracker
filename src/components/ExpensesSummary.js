import React from 'react';
import {connect} from 'react-redux';
import numeral from 'numeral';
import selectExpenses from '../selectors/expenses';
import selectTotalExpenses from '../selectors/expenses-total';

 export const ExpensesSummary = props => (
    <div className="alert alert-dismissible alert-info">
        {
            props.expensesCount === 1 
                ? <strong>There is {props.expensesCount} expense, with a total of {numeral(props.expensesTotal / 100).format('$0,0.00')}</strong>
                : <strong>There are {props.expensesCount} expenses, with a total of {numeral(props.expensesTotal / 100).format('$0,0.00')}</strong>
        }
    </div>
);

const mapStateToProps = state => ({
    expensesCount: selectExpenses(state.expenses, state.filters).length,
    expensesTotal: selectTotalExpenses(selectExpenses(state.expenses, state.filters))
});

export default connect(mapStateToProps)(ExpensesSummary);