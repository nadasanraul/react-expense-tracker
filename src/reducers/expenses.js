//Reducer default state
const expensesReducerDefaultState = []

// Reducer function
export default (state = expensesReducerDefaultState, action) => {
    switch(action.type){
        case 'ADD_EXPENSE':
            return [...state, action.payload.expense];
        case 'REMOVE_EXPENSE':
            return state.filter(({id}) => id !== action.payload.id);
        case 'EDIT_EXPENSE':
            return state.map(expense => expense.id === action.payload.id ? {...expense, ...action.payload.updates} : expense); 
        case 'SET_EXPENSES':
            return action.payload.expenses
        default:
            return state;
    }
};
