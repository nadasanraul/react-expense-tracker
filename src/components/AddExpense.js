import React from 'react';
import {connect} from 'react-redux';
import ExpenseForm from './ExpenseForm';
import {addExpense} from '../actions/expenses';

export class AddExpense extends React.Component{
    onSubmit = expense => 
        //dispatch(addExpense(expense));
        this.props.addExpense(expense);
    
    render() {
        return (
            <div className="container">
                <h1>Add Expense</h1>
                <ExpenseForm 
                    onSubmit={this.onSubmit}
                />
            </div>
        );
    }
}


const mapDispatchToProps = (dispatch) => ({
    addExpense: expense => dispatch(addExpense(expense))
});


export default connect(undefined, mapDispatchToProps)(AddExpense);