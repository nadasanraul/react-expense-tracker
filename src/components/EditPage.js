import React from 'react';
import {connect} from 'react-redux';
import ExpenseForm from './ExpenseForm';
import {editExpense, startRemoveExpense} from '../actions/expenses';

export class EditPage extends React.Component {

    onSubmit = expense => {
        this.props.editExpense(this.props.expense.id, expense)};

    onRemove = () => {
        this.props.startRemoveExpense({id: this.props.expense.id});
        this.props.history.push('/');
    }

    render () {
        return (
            <div className="container">
                <ExpenseForm 
                    expense={this.props.expense}
                    onSubmit={this.onSubmit}
                />
                <button className="btn btn-danger" onClick={
                    this.onRemove}
                >
                    Remove
                </button>
            </div>
        )
    }
}

// const EditPage = (props) =>  (
//     <div className="container">
//         <ExpenseForm 
//             expense={props.expense}
//             onSubmit={expense => {
//                 props.dispatch(editExpense(props.expense.id, expense));
//             }}
//         />
//         <button className="btn btn-danger" onClick={(e) => {
//             props.dispatch(removeExpense({id: props.expense.id}));
//             props.history.push('/');
//         }}>Remove</button>
//     </div>
// );

const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find(expense => expense.id === props.match.params.id)
    }
};

const mapDispatchToProps = (dispatch) => ({
    editExpense: (id,expense) => dispatch(editExpense(id, expense)),
    startRemoveExpense: ({id}) => dispatch(startRemoveExpense({id}))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditPage);