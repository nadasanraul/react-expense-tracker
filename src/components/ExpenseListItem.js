import React from 'react';
import {Link} from 'react-router-dom';
import {connect} from 'react-redux';
import moment from 'moment';
import numeral from 'numeral';
import {startRemoveExpense} from '../actions/expenses';

export const ExpenseListItem = ({id, description, amount, createdAt, dispatch}) => (
    <div className="card border-primary mb-3">
        <div className="card-body">
            <h2 className="card-title">{description}</h2>
            <p className="card-text">Amount:{numeral(amount / 100).format('$0,0.00')} </p>
            <p className="card-text">On: {moment(createdAt).format('Do MMM, YYYY')}</p>
            <button className="btn btn-danger" onClick={(e) => {
                dispatch(startRemoveExpense({id}));
            }}>Remove</button>
            <Link to={`edit/${id}`} className="btn btn-info float-right">Edit</Link>
        </div>
    </div>
);

export default connect()(ExpenseListItem);