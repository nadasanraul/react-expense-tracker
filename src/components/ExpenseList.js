import React from 'react';
import {connect} from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

export const ExpenseList = (props) => (
    <div>
        {
            props.expenses.length === 0 ? (
                <h2 className="text-center">No expenses</h2>
            ) : (
                props.expenses.map(expense => 
                    <ExpenseListItem key={expense.id} {...expense}/>
                )
            )
        }
    </div>
);

const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    } 
};

export default connect(mapStateToProps)(ExpenseList);
