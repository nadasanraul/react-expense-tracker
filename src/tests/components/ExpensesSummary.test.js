import React from 'react';
import {shallow} from 'enzyme';
import {ExpensesSummary} from '../../components/ExpensesSummary';
import selectTotalExpenses from '../../selectors/expenses-total';
import expenses from '../fixtures/testExpenses';

let expense, wrapper;

beforeEach(() => {
    expense = [expenses[1]];
    wrapper = shallow(<ExpensesSummary expensesCount={expenses.length} expensesTotal={selectTotalExpenses(expenses)}/>)
});

test('should render ExpensesSummary correctly with 1 item', () => {
    wrapper.setProps({
        expensesCount: expense.length,
        expensesTotal: selectTotalExpenses(expense)
    });
    expect(wrapper).toMatchSnapshot();
});

test('should render ExpensesSummary correctly with multiple or 0 items', () => {
    expect(wrapper).toMatchSnapshot();
});