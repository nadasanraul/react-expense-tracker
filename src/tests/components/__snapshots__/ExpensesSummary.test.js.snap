// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpensesSummary correctly with 1 item 1`] = `
<div
  className="alert alert-dismissible alert-info"
>
  <strong>
    There is 
    1
     expense, with a total of 
    $1,090.05
  </strong>
</div>
`;

exports[`should render ExpensesSummary correctly with multiple or 0 items 1`] = `
<div
  className="alert alert-dismissible alert-info"
>
  <strong>
    There are 
    3
     expenses, with a total of 
    $1,137.00
  </strong>
</div>
`;
